### ISSUE TITLE
Polygon errors when using perPositionHeight

### ISSUE THREAD

**Issue by ChriWe** [AnalyticalGraphicsInc/cesium#5203]
My goal is the visualization of an **elevated extruded polygon** with data provided to a `GeoJsonDataSource`.
- The polygon is given by a GeoJson dataset, consisting of points where each is given by longitude, latitude and elevation (e.g. [ 30.0, 70.0, 152452.62 ])
- The polygon should be extruded by 30502.

A schematic profile view of such a volume could look like this:
- Green: Bottom of the polygon (given by the dataset) -> should follow the ellipsoid in 3D
- Red: Extrusion
- Orange: Top of the polygon (given by dataset & extrusion) -> should follow the ellipsoid in 3D
![profile_view](https://cloud.githubusercontent.com/assets/11804771/24970531/f09651c4-1fb4-11e7-9a9e-8a5e3fd74576.png)

**Now to the pratical part:** I am using the following GeoJson data: [polygon.zip](https://github.com/AnalyticalGraphicsInc/cesium/files/917214/polygon.zip)

When using the data provided above, the visualization with `GeoJsonDataSource` shows some strange rendering effects. It looks like the geometry folds and cuts the globe (bottom and top of the polygon not properly following the ellipsoid):
![polygon_rendering](https://cloud.githubusercontent.com/assets/11804771/24967755/6f784ff6-1fab-11e7-8c96-1cb4d7ed402a.png)

I have also tried to visualize the polygon by using `PolygonGeometry` but unfortunately there are also some rendering issues:
- There is an edge in the visualization (the geometry seems to fold or cut the earth)
![polygongeometry](https://cloud.githubusercontent.com/assets/11804771/24967931/e906a624-1fab-11e7-979b-d1f4976cfd05.png)
- When having certain view-angles the folding/cut is even more visible
![polygongeometry_rendering](https://cloud.githubusercontent.com/assets/11804771/24968025/2ce2799a-1fac-11e7-8246-4bca7dd42ac2.png)

So basically the issue is, that the top and bottom surface of the volume do not follow the ellipsoid properly.
- Is there any way to avoid those rendering artifacts? 
- Is this issue known? Especially when using very large polygons?
Actually I'm not quite sure if this is a bug, I'm missing some configurations or I´m using a wrong approach. I also could not find any sample yet which is dealing with such kind of visualizations.

I would really appreciate any help. :)

**Comment by hpinkos** [comment]
Thanks for including all of the details and screenshots @ChriWe!
The short answer is that Cesium currently doesn't have support for drawing this kind of polygon.

There are a few known issues with using the `perPositionHeight` property of `PolygonGeometry` that are resulting in your polygon being visualized this way.  The first is that the polygon is not being subdivided to fit the curvature of the earth.  That's why it looks like it's being cut off by the globe: for really large polygons it's actually slicing through the globe.

The second issue is that due to our current triangulation algorithm, `perPositionHeight` works best for polygons with coplanar points.  Otherwise, sometimes the shape gets divided in a way that differs from what you expect.  For example, imagine a shape that's like the roof of a house.  Here's a drawing that represents two different ways that the points could be split into triangles.  The first way is what we'd expect to happen, but sometimes the second one happens and as a result the shape draws differently than we want it to.

![image](https://cloud.githubusercontent.com/assets/3451886/24972346/28042a0e-1f89-11e7-8eb3-df7f3994ce72.png)

**Comment by hpinkos** [comment]
Related #2399

**Comment by ChriWe** [comment]
Thanks very much for the fast reply. Also for the drawings ;).

Is the implementation of such a feature planned in the near future?

For the `GeoJsonDataSource` example `perPositionHeight` was set to false.
For the ` PolygonGeometry` example `perPositionHeight` was set to true.
So I am not sure if the inproper visualization of example 1 is due to the `perPositionHeight` property. But yeah, looks like a triangulation issue.

Could you imagine any workaround for this? A potential one I can see is the consolidation of the given dataset (e.g. Delauney Triangulation) with a grid-spacing depending on earth´s curvature. This could be used as a base for visualization. Maybe via comosition of multiple polygons. Seems like a bigger pile of work. But pls correct me if I´m wrong. 

BTW: What really confused me at first was the different visualization of polygons in connection with the `GeoJsonDataSource`:
- If elevation data is provided in the GeoJson, then it´s flat (and also slices the earth if big enough).
- If there is just longitude & latitude, it properly follows the curvature of the earth.
Also could not find this behavior in the docs (neither GeoJsonDataSource nor PolygonGraphics).

**Comment by hpinkos** [comment]
I'm not sure when we'll have a chance to look into fixing this.  We're focused on finishing up the 3d tiles specification right now, so this would only be a priority if it ties into our vector tiles implementation.

`perPositionHeight` is for sure the culprit.  If you look in the code for `GeoJsonDataSource`, it tells the polygons to [use `perPositionHeight` when the GeoJSON coordinates have a height](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/DataSources/GeoJsonDataSource.js#L433-L434 ). 
That's why it only curves with the earth when there is no elevation data.  We subdivide the polygon correctly when `perPositionHeight` is not used.

We would be happy to review a pull request if you wanted to take a stab at fixing this!  You would have to pass the `perPositionHeight` variable through here [PolygonGeometryLibrary.createGeometryFromPositions](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/PolygonGeometryLibrary.js#L301)
Then use a different triangulation algorithm if `perPositionHeight === true` [PolygonPipeline.triangulate ](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/PolygonPipeline.js#L83-L92)

For the curvature of the earth problem, you would need to make [PolygonPipeline.computeSubdivision](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/PolygonPipeline.js#L114) work with a polygon with `perPositonHeight`.  We currently [skip the subdivision](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Core/PolygonGeometryLibrary.js#L310) if `perPositionHeight` is being used.

See our documentation for information on [building the code base](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/Documentation/Contributors/BuildGuide) and [opening a pull request](https://github.com/AnalyticalGraphicsInc/cesium/blob/master/CONTRIBUTING.md#opening-a-pull-request)

**Comment by ChriWe** [comment]
Ok, thanks for the info.

Actually this feature would be a nice-to-have in the current application im building, but wasn't yet put on the roadmap. So I may have a look at this at some future point.

Thank you very much for the detailed information and description.

**Comment by ggetz** [comment]
I think this problem experienced on the forum is related: https://groups.google.com/forum/#!searchin/cesium-dev/paula%7Csort:relevance/cesium-dev/FiF8tbgoPjw/ZXHBhlhABAAJ

##### With `perPositionHeight = false`
![image](https://user-images.githubusercontent.com/4439461/29685134-83190a8c-88e2-11e7-98bf-aaf6354ce77f.png)

##### With `perPositionHeight = true`
![image](https://user-images.githubusercontent.com/4439461/29685139-86864450-88e2-11e7-8aa9-4367f705d192.png)

**Comment by pjcozzi** [comment]
@ggetz this issue is closed, do you want to re-open it or create a new one?

**Comment by ggetz** [comment]
This seems like the same problem, so reopen please.

**Comment by hpinkos** [comment]
The above looks like a data problem.  I think his position heights are lower than the extrudedHeight.  I'll add a comment to #2924

**Comment by cesium-concierge** [comment]
Congratulations on closing the issue! I found these Cesium forum links in the comments above:

https://groups.google.com/forum/#!searchin/cesium-dev/paula%7Csort

If this issue affects any of these threads, please post a comment like the following:

> The issue at https://github.com/AnalyticalGraphicsInc/cesium/issues/5203 has just been closed and may resolve your issue. Look for the change in the next stable release of Cesium or get it now in the master branch on GitHub https://github.com/AnalyticalGraphicsInc/cesium.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:
